tracer-transformers/src/Control/Tracer/Transformers/Synopsizer.lhs

```diff
@@ -0,0 +1,99 @@
+\label{code:Control.Tracer.Transformers.WithThreadAndTime}
+
+%if style == newcode
+\begin{code}
+{-|
+Module: Synopsizer
+
+Synopsize runs of repeated events, as an initial message with a summary at the end.
+-}
+{-# LANGUAGE BangPatterns   #-}
+{-# LANGUAGE LambdaCase     #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE MultiWayIf     #-}
+{-# LANGUAGE RankNTypes     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Control.Tracer.Transformers.Synopsizer
+    (
+    -- * transformer
+      mkSynopsizer
+    , Synopsized(..)
+    ) where
+
+import           Control.Monad.IO.Class (MonadIO (..))
+import           Data.IORef (newIORef,readIORef, writeIORef)
+
+import           Control.Tracer (Tracer (..), traceWith)
+
+\end{code}
+%endif
+
+\begin{code}
+-- | State required to track and summarise repeats.
+data SynopsizerState a
+  = SynopsizerState
+    { ssRepeats  :: {-# UNPACK #-} !Int
+    , ssLast     :: !(Maybe a)
+    }
+
+-- | Abstract representation of a message stream with synopses.
+data Synopsized a
+  = One !a
+    -- ^ Normal, unsynopsized message.
+  | Many {-# UNPACK #-} !Int !a
+    -- ^ Synopsis for a specified number of messages similar to a previous one.
+    --   Note that several 'Many' messages can follow the initial 'One' message,
+    --   because of configurable synopsis overflow.
+
+-- | Generic Tracer transformer.  Will omit consequent runs of similar log objects,
+--   with similarity defined by a supplied predicate.  A run of similar log
+--   objects is terminated by one of two events:
+--     1. a dissimilar event, or
+--     2. reaching the specified overflow limit for omitted messages.
+--
+--   At the end of a run of similar objects, a summary of omission is emitted.
+--
+--   Caveat 1:  the supplied criterion is expected to define a transitive relation.
+--
+--   Caveat 2:  the resulting tracer has state, which is lost upon trace termination.
+--   This means that if the trace ends with a run of similar messages, this will
+--   not be reflected in the trace itself, as observed by the backends
+--   -- they'll only receive the first message of the run.
+--
+--   WARNING:  the resulting tracer is not thread-safe!
```

Comment:
Sorry - why can't you use modifyIORef (strict or not) here? 
You could return the resulting tracing actions an then execute them outside the critical region.
This is relatively easy to make thread safe.

Reply:
Alternatively you could use an MVar - which, combined with, withMVar (or variant) will get you the desired properties.

Reply:
@njd42, we've had a fairly detailed discussion with @CodiePP regarding the tradeoffs. and the gist of it is that:

1. We don't want to incur additional locking overhead on basically all users.
2. There are no known cases where a tracer is shared across threads.